merkletree.go

```diff
@@ -0,0 +1,242 @@
+package merkletree
+
+import "errors"
+
+var (
+	ErrInvalidTree = errors.New("[merkletree] invalid tree")
+)
+
+const (
+	EmptyBranchIdentifier = 'E'
+	LeafIdentifier        = 'L'
+)
+
+type MerkleTree struct {
+	treeNonce []byte
+	salt      []byte
+	privKey   []byte
+	pubKey    []byte
+	root      *interiorNode
+}
+
+func InitMerkleTree(treeNonce, salt, pubKey, privKey []byte) *MerkleTree {
+	root := &interiorNode{
+		parent:     nil,
+		leftChild:  new(emptyNode),
+		rightChild: new(emptyNode),
+		leftHash:   nil,
+		rightHash:  nil,
+		level:      0,
+	}
+
+	m := &MerkleTree{
+		treeNonce: treeNonce,
+		salt:      salt,
+		root:      root,
+		privKey:   privKey,
+		pubKey:    pubKey,
+	}
+	return m
+}
+
+func LookUp(key string) (MerkleNode, []MerkleNode, error) {
+	str := getCurrentSTR()
+	return lookUp(key, str)
+}
+
+func LookUpInEpoch(key string, ep int64) (MerkleNode, []MerkleNode, error) {
+	str := GetSTR(ep)
+	return lookUp(key, str)
+}
+
+func lookUp(key string, str *SignedTreeRoot) (MerkleNode, []MerkleNode, error) {
+	lookupIndex := computePrivateIndex(key)
+	position := 0
+	var nodePointer interface{}
+	nodePointer = str.treeRoot
+	var proof []MerkleNode
+
+	for {
+		if _, ok := nodePointer.(*userLeafNode); ok {
+			// reached to a leaf node
+			break
+		}
+		if _, ok := nodePointer.(*emptyNode); ok {
+			// reached to an empty branch
+			break
+		}
+		proof = append(proof, nodePointer.(*interiorNode))
+		direction := getNthBit(lookupIndex, position)
+		if direction {
+			nodePointer = nodePointer.(*interiorNode).rightChild
+		} else {
+			nodePointer = nodePointer.(*interiorNode).leftChild
+		}
+		position++
+	}
+
+	if nodePointer == nil {
+		return nil, nil, ErrInvalidTree
+	}
+	switch nodePointer.(type) {
+	case *userLeafNode:
+		if nodePointer.(*userLeafNode).key == key {
+			proof = append(proof, nodePointer.(*userLeafNode))
+			return nodePointer.(*userLeafNode), proof, nil
+		} else {
+			// TODO: what should we return here?
```

Comment:
@masomel : in case the lookup prefix is same as the prefix of an index (i.e. an user leaf node) in the tree, but they are actually different, what should we return for proof of absence?